### PR TITLE
[Pallas] Add Pallas design doc

### PR DESCRIPTION
**Pull request by sharadmv**

### CODE REVIEW

File: docs/pallas/design.md [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+
+Writing to `Ref`s can be done via analogous `__setitem__` style indexing.
+
+Other forms of indexing (for example, dynamic slicing) can be done via `pallas.load` and `pallas.store`, new JAX primitives designed to make loading from/storing into memory easier. We’ll discuss these new primitives later.
+
+### Extending JAX with new Pallas primitives
+
+Because JAX was designed with HLO in mind, the set of JAX primitives closely mirrors the set of HLO operations. Targeting a new compiler (e.g. Triton or Mosaic) means we might need to supplement JAX’s primitives with new ones specific to the new compiler. At the same time, we may not be able to lower all JAX primitives, so we need to restrict it to a subset.
+
+Because Pallas was initially designed with Triton in mind, we offer a set of new primitives with the Triton programming model in mind. As we’ll show later, we can lower these primitives to Mosaic as well.
```
Comment (jakevdp) [inline docs/pallas/design.md:88]: word repetition:
```suggestion
Because Pallas was initially designed with Triton in mind, we offer a set of new primitives targeting the Triton programming model. As we’ll show later, we can lower these primitives to Mosaic as well.
```

File: docs/pallas/design.md [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+
+
+`pallas.load` and `pallas.store` also support masking via the mask argument.
+
+```python
+def f(x_ref, o_ref):
+  # Reading from memory via pallas.load
+  idx = jnp.arange(8)
+  mask = idx < 5
+  x = pl.load(x_ref, (idx,), mask=mask, other=float('-inf'))
```
Comment (jakevdp) [inline docs/pallas/design.md:111]: Have you ever considered adopting JAX-style `at[].get()` syntax for this? e.g. you could imagine writing this `pl.load` statement as
```python
x = x_ref.at[idx].get(mask=mask, other=float('-inf'))
```

File: docs/pallas/design.md [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+
+### Extending JAX with new Pallas primitives
+
+Because JAX was designed with HLO in mind, the set of JAX primitives closely mirrors the set of HLO operations. Targeting a new compiler (e.g. Triton or Mosaic) means we might need to supplement JAX’s primitives with new ones specific to the new compiler. At the same time, we may not be able to lower all JAX primitives, so we need to restrict it to a subset.
+
+Because Pallas was initially designed with Triton in mind, we offer a set of new primitives with the Triton programming model in mind. As we’ll show later, we can lower these primitives to Mosaic as well.
+
+#### `pallas.load` and `pallas.store`
+`pallas.load` and `pallas.store` are primitives that allow loading from memory and storing into memory. Unlike `__getitem__` and `__setitem__` they are more flexible at the cost of being more verbose. Specifically, you can use the `pallas.dynamic_slice` (`pallas.ds` for short) construct (which should maybe be upstreamed into JAX to be used with Ref `__getitem_` and `__setitem__`).
```
Comment (jakevdp) [inline docs/pallas/design.md:91]:
```suggestion
`pallas.load` and `pallas.store` are primitives that allow loading from memory and storing into memory. Unlike `__getitem__` and `__setitem__` they are more flexible at the cost of being more verbose. Specifically, you can use the `pallas.dynamic_slice` (`pallas.ds` for short) construct (which should maybe be upstreamed into JAX to be used with Ref `__getitem__` and `__setitem__`).
```

File: docs/pallas/design.md (hunk 2 above)
Comment (sharadmv) [reply docs/pallas/design.md:111]: We could do something like that, though I wanted to draw a contrast between slicing an array and loading from a `Ref`, since `Ref`s are not arrays. Maybe something like `x_ref.at[idx].load(...)` and `x_ref.at[idx].store(...)`.

**Review by jakevdp — APPROVED** [review]

**Review by mattjj — APPROVED** [review]
Looks great!

File: docs/pallas/design.md (hunk 2 above)
Comment (sharadmv) [reply docs/pallas/design.md:111]: Yea I think that's a good idea @froystig @mattjj

Comment (mattjj) [reply docs/pallas/design.md:111]: Can you summarize the upside, e.g. with an example?

Comment (mattjj) [reply docs/pallas/design.md:111]: IIUC it's roughly that we'd replace `lax.dynamic_slice(x, [idx], [size])` with `x[ds(idx, size)]`?
